jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java

```diff
@@ -68,56 +103,152 @@ static RetainableByteBuffer wrap(ByteBuffer byteBuffer)
+                return released;
             }
+        };
+    }
 
-            @Override
-            public boolean isRetained()
-            {
-                throw new UnsupportedOperationException();
-            }
+    /**
+     * Check if the underlying implementation is mutable.
+     * Note that the immutable {@link RetainableByteBuffer} API may be backed by a mutable {@link ByteBuffer} or
+     * the {@link Mutable} API may be backed by an immutable {@link ByteBuffer}.
+     * @return whether this buffers implementation is mutable
+     * @see #asMutable()
+     */
+    default boolean isMutable()
+    {
+        return !getByteBuffer().isReadOnly();
+    }
 
-            @Override
-            public boolean canRetain()
-            {
-                return retainable.canRetain();
-            }
+    /**
+     * Access this buffer via the {@link Mutable} API.
+     * Note that the {@link Mutable} API may be backed by an immutable {@link ByteBuffer}.
+     * @return An {@link Mutable} representation of this buffer with same data and pointers.
+     * @throws ReadOnlyBufferException If the buffer is not {@link Mutable} or the backing {@link ByteBuffer} is
+     * {@link ByteBuffer#isReadOnly() read-only}.
+     * @see #isMutable()
+     */
+    default Mutable asMutable() throws ReadOnlyBufferException
+    {
+        if (!isMutable() || isRetained())
+            throw new ReadOnlyBufferException();
+        if (this instanceof Mutable mutable)
+            return mutable;
+        throw new ReadOnlyBufferException();
+    }
 
-            @Override
-            public void retain()
-            {
-                retainable.retain();
-            }
+    /**
+     * Appends and consumes the contents of this buffer to the passed buffer, limited by the capacity of the target buffer.
+     * @param buffer The buffer to append bytes to, whose limit will be updated.
+     * @return {@code true} if all bytes in this buffer are able to be appended.
+     * @see #putTo(ByteBuffer)
+     */
+    default boolean appendTo(ByteBuffer buffer)
+    {
+        return remaining() == BufferUtil.append(buffer, getByteBuffer());
+    }
 
-            @Override
-            public boolean release()
-            {
-                return retainable.release();
-            }
-        };
+    /**
+     * Appends and consumes the contents of this buffer to the passed buffer, limited by the capacity of the target buffer.
+     * @param buffer The buffer to append bytes to, whose limit will be updated.
+     * @return {@code true} if all bytes in this buffer are able to be appended.
+     * @see #putTo(ByteBuffer)
+     */
+    default boolean appendTo(RetainableByteBuffer buffer)
+    {
+        return appendTo(buffer.getByteBuffer());
+    }
+
+    /**
+     * Creates a deep copy of this RetainableByteBuffer that is entirely independent
+     * @return A copy of this RetainableByteBuffer
+     */
+    default RetainableByteBuffer copy()
```

Comment:
I'm wondering if that API isn't going to cause more harm than good as I fail to see when we'd like to make copies into newly allocated buffers, except in tests.

Reply:
It is used in the take implementations.... which themselves are mostly used for tests..... but copy is something likely to be needed if usage of these APIs becomes more widespread, so I'd much rather we implement it and test it correctly.

If we don't use it, then it is not a problem.  If we do, then it is needed.